### PR TITLE
[Merged by Bors] - feat(RingTheory): composition of presentations

### PR DESCRIPTION
**Pull request by chrisflav**
Given an `R`-algebra `S` and an `S`-algebra `T` with presentations `P` and `Q` we construct the canonical `R`-presentation of `T`.

This contribution was created as part of the AIM workshop "Formalizing algebraic geometry"
in June 2024.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/Presentation.lean [hunk 1]
```diff
@@ -229,6 +230,121 @@ def baseChange : Presentation T (T ⊗[R] S) where
...
+
+variable {T} [CommRing T] [Algebra S T]
+variable (Q : Presentation S T) (P : Presentation R S)
+
+private noncomputable def comp_relation_aux (r : Q.rels) : MvPolynomial (Q.vars ⊕ P.vars) R :=
+  Finsupp.sum (Q.relation r)
+    (fun x j ↦ (MvPolynomial.rename Sum.inr <| P.σ j) * monomial (x.mapDomain Sum.inl) 1)
+
+private noncomputable def aux : MvPolynomial (Q.vars ⊕ P.vars) R →ₐ[R] MvPolynomial Q.vars S :=
+  aeval (Sum.elim X (MvPolynomial.C ∘ P.val))
```
Comment (jcommelin) [inline Mathlib/RingTheory/Presentation.lean:245]: Please write a module docstring at the beginning of this section, explaining the goal and strategy. And please explain the role of `aux` (and maybe give it a docstring).

Comment (chrisflav) [reply Mathlib/RingTheory/Presentation.lean]: Done, if it is not clear yet, please complain.

**Review by jcommelin — COMMENTED** [review]
maintainer merge

File: Mathlib/RingTheory/Ideal/Maps.lean [hunk 2]
```diff
@@ -125,11 +125,21 @@ theorem comap_comap {T : Type*} [Semiring T] {I : Ideal T} (f : R →+* S) (g :
     (I.comap g).comap f = I.comap (g.comp f) :=
   rfl
 
+lemma comap_comapₐ {R A B C : Type*} [CommSemiring R] [Semiring A] [Algebra R A] [Semiring B]
```
Comment (riccardobrasca) [inline Mathlib/RingTheory/Ideal/Maps.lean:128]: Isn't `R` already in scope?

File: Mathlib/RingTheory/Presentation.lean [hunk 3]
```diff
@@ -266,6 +267,161 @@ instance baseChange_isFinite [P.IsFinite] : (P.baseChange T).IsFinite where
...
+
+/-- The pre-images constructed in `comp_relation_aux` are indeed pre-images under `aux`. -/
+private lemma comp_relation_aux_map (r : Q.rels) :
+    (Q.aux P) (Q.comp_relation_aux P r) = Q.relation r := by
+  simp only [aux, comp_relation_aux, Generators.comp_vars, Sum.elim_inl, map_finsupp_sum]
+  simp only [_root_.map_mul, aeval_rename, aeval_monomial, Sum.elim_comp_inr]
+  conv_rhs => rw [← Finsupp.sum_single (Q.relation r)]
+  congr
+  ext u s m
+  show _ = coeff m (monomial u s)
```
Comment (riccardobrasca) [inline Mathlib/RingTheory/Presentation.lean:335]: Is there a `rfl` lemma we can rewrite, instead of using `show`?

**Review by riccardobrasca — APPROVED** [review]
Thanks!

bors d+

File: Mathlib/RingTheory/Ideal/Maps.lean (hunk 2 above)
Comment (chrisflav) [reply Mathlib/RingTheory/Ideal/Maps.lean:128]: Yes, but it is only a `Semiring`, but I need `CommSemiring` for `Algebra R A` to make sense.

File: Mathlib/RingTheory/Presentation.lean (hunk 3 above)
Comment (chrisflav) [reply Mathlib/RingTheory/Presentation.lean]: Yes, indeed, thanks!

**Comment by chrisflav** [comment]
Thanks for the reviews!

**Pull request closed without merging**